aten/src/ATen/native/cuda/SpectralOps.cu

```diff
@@ -0,0 +1,274 @@
+  long long int batch = input.size(0);
+
+  // check the input sizes and strides to see if we need to make it contiguous
+  std::vector<long long int> signal_sizes(signal_ndim);
+  std::vector<long long int> inembeds(signal_ndim);
+  long long int istride, signal_size;
+  bool need_contiguous = input.stride(signal_ndim) == 0;
+  if (complex_input) {
+    // real/imag dimension must be like complex type
+    need_contiguous |= input.stride(-1) != 1;
+  } else if (is_half) {
+    // For half, base strides on the real part of real-to-complex and
+    // complex-to-real transforms are not supported. Since our output is always
+    // contiguous, only need to check real-to-complex case.
+    need_contiguous |= input.stride(signal_ndim) != 1;
+  }
+  // store last tensor stride to infer inembeds array
+  // complex input's last dim is size 2 and contiguous
+  long long int ilast_stride = complex_input ? 2 : 1;
+   // for each signal dim from innermost to outermost
+  for (int64_t i = signal_ndim - 1; i >= 0; i--) {
+    signal_size = checked_signal_sizes[i];
+    istride = input.stride(i + 1);
+    if (is_half && !is_pow_of_two(signal_size)) {
+      std::ostringstream ss;
+      ss << "cuFFT doesn't support signals of half type with size that is not "
+         << "a power of two, but gets a signal size of " << signal_size << " at"
+         << "signal dimension " << i;
+      throw std::runtime_error(ss.str());
+    }
+    signal_sizes[i] = signal_size;
+    if (!need_contiguous) {
+      // set the inembeds for dim in last dimension
+      if (ilast_stride == 0) {
+        need_contiguous = istride != 0;
+        if (i < signal_ndim - 1) {
+          inembeds[i + 1] = 0;
+        }
+      } else {
+        need_contiguous = istride % ilast_stride != 0;
+        if (i < signal_ndim - 1) {
+          inembeds[i + 1] = istride / ilast_stride;
+        }
+      }
+      ilast_stride = istride;
+    }
+  }
+
+  need_contiguous |= (input.size(0) > 1 && input.stride(0) == 0);
+  if (need_contiguous) {
+    input = input.contiguous();
+    auto input_sizes = input.sizes();
+    std::copy(&input_sizes[1], &input_sizes[signal_ndim] + 1, &inembeds[0]);
+  }
+
+  // set base_istride (stride at innermost dim of signal)
+  // set idist (stride at batch dim)
+  long long int idist = complex_input ? input.stride(0) >> 1 : input.stride(0);
+  // Even if batch dimension is one, then idist (stride(0)) doesn't matter,
+  // cuFFT errors if idist = 0. This is hack to make it succeed.
+  if (idist == 0 && batch == 1) {
+    idist = 1;
+  }
+  long long int base_istride = complex_input ? input.stride(signal_ndim) >> 1 : input.stride(signal_ndim);
+
+  // output, base_ostride, odist
+  auto output = input.type().tensor(output_sizes);
+  std::vector<long long int> onembeds(&output_sizes[1], &output_sizes[signal_ndim] + 1);
+  long long int odist = complex_output ? output.stride(0) >> 1 : output.stride(0);
+  long long int base_ostride = 1;
+
+  CufftHandle plan;
```

Comment:
cufft seems to have its own streaming mechanism (`cufftSetStream`). The docs state:

>  If no stream is associated with a plan, launches take place in stream(0), the default CUDA stream.

Do you have to do something about this? Are you testing fft on CUDA tensors which are on a non-default device?